### ISSUE TITLE
Thread unsafety

### ISSUE THREAD

**Issue by baszalmstra** [hukumka/generic_static#4]
The code is *not* thread safe. The reader writer lock introduces subtle bugs. 

This can happen:

| thread 1           | thread 2           |
|--------------------|--------------------|
| call_once<T>       | call_once<T>       |
| acquire read lock  | acquire read lock  |
| found missing      | found missing      |
| call f()           | call f()           |
| acquire write lock | ..                 |
| store value        | ..                 |
| ..                 | acquire write lock |
| ..                 | store value        |
|                    | PANIC              |

This can be fixed by rechecking whether the value has been inserted when the write lock is acquired. However `f` must be called outside the write lock because otherwise you could get a deadlock if `f` recursively calls `call_once<S>`.

**Comment by hukumka** [comment]
Thanks for the report!
Have recent changes fixed the issue?
(05115734a8626081cd459356966be0beb4a5ba24 - 448e4a9505696c06120da7e861d58d386a03af97)

**Comment by baszalmstra** [comment]
That will work yes, but you have to consider that the function f could now
by called more than once (since its no longer guarded by a lock), so that
might not be the correct name anymore and could potentially have other
consequences.

**Comment by hukumka** [comment]
Good catch, I totally missed it. I think it could be solved by wrapping each entry in OnceCell.

**Comment by hukumka** [comment]
80af1344aabcb8b02d7b86c10f9380af3b53f9af should do the trick.

**Comment by baszalmstra** [comment]
I think this can still result in a deadlock because the writer still holds the lock when calling f().

**Comment by hukumka** [comment]
Yes, f also should not be called inside reader lock
